{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5f49ebb8-4600-435e-9ea6-6037dd09b4ec",
   "metadata": {
    "tags": []
   },
   "source": [
    "# **2D (XY) Clustergram** with Python-Plotly interactive graphing <br> *using the gene expression dataset*\n",
    "## Learn more about Clastergram Chart: \n",
    "- https://plotly.com/python/clustergram/\n",
    "- https://dash.plotly.com/dash-bio/clustergram\n",
    "\n",
    "*To run all cells in this notebook, please use `Run`, then `Run All Cells` options in the top menu. Alternatively, you can press â–¶ button above or `ctrl + enter/return` on each cell separately. To modify the content of the cell:*\n",
    "* *use single-click on the code-like cell*\n",
    "* *use double-click on the text-like cell*\n",
    "\n",
    "Color variables in **Plotly** have a string type value provided in the one of the following forms:\n",
    "- HEX color, e.g., \"#FFFFFF\"\n",
    "- HSL color, e.g., \"hsl(hue, saturation,lightness)\"; The Hue values are in degrees, from 0 to 360 (0 - red, 120 - green, 240 - blue). The saturation and lightness are in % from 0% to 100%.\n",
    "- RGB color, e.g., \"rgb(red, green, blue)\"; Each channel is provided as an intiger between 0 and 255, e.g., rgb(128, 0, 128).\n",
    "- RGBA color, e.g., \"rgba(r, g, b, a)\" with 'a' value set up in range from 0.0 (transparent) to 1.0 (solid)\n",
    "- common name, e.g., \"blue\"\n",
    "\n",
    "**Built-in Plotly colors** *(common names)*\n",
    "\n",
    "<span style=\"background-color:aliceblue\">aliceblue</span>, \n",
    "<span style=\"background-color:antiquewhite\">antiquewhite</span>, \n",
    "<span style=\"background-color:aqua\">aqua</span>, \n",
    "<span style=\"background-color:aquamarine\">aquamarine</span>, \n",
    "<span style=\"background-color:azure\">azure</span>, \n",
    "<span style=\"background-color:beige\">beige</span>, \n",
    "<span style=\"background-color:bisque\">bisque</span>, \n",
    "<span style=\"background-color:black; color: white\">black</span>, \n",
    "<span style=\"background-color:blanchedalmond\">blanchedalmond</span>, \n",
    "<span style=\"background-color:blue\">blue</span>,\n",
    "<span style=\"background-color:blueviolet\">blueviolet</span>, \n",
    "<span style=\"background-color:brown\">brown</span>, \n",
    "<span style=\"background-color:burlywood\">burlywood</span>, \n",
    "<span style=\"background-color:cadetblue\">cadetblue</span>, \n",
    "<span style=\"background-color:chartreuse\">chartreuse</span>, \n",
    "<span style=\"background-color:chocolate\">chocolate</span>, \n",
    "<span style=\"background-color:coral\">coral</span>, \n",
    "<span style=\"background-color:cornflowerblue\">cornflowerblue</span>, \n",
    "<span style=\"background-color:cornsilk\">cornsilk</span>, \n",
    "<span style=\"background-color:crimson\">crimson</span>, \n",
    "<span style=\"background-color:cyan\">cyan</span>, \n",
    "<span style=\"background-color:darkblue\">darkblue</span>, \n",
    "<span style=\"background-color:darkcyan\">darkcyan</span>, \n",
    "<span style=\"background-color:darkgoldenrod\">darkgoldenrod</span>, \n",
    "<span style=\"background-color:darkgray\">darkgray</span>, \n",
    "<span style=\"background-color:darkgrey\">darkgrey</span>, \n",
    "<span style=\"background-color:darkgreen\">darkgreen</span>, \n",
    "<span style=\"background-color:darkkhaki\">darkkhaki</span>, \n",
    "<span style=\"background-color:darkmagenta\">darkmagenta</span>, \n",
    "<span style=\"background-color:darkolivegreen\">darkolivegreen</span>, \n",
    "<span style=\"background-color:darkorange\">darkorange</span>, \n",
    "<span style=\"background-color:darkorchid\">darkorchid</span>, \n",
    "<span style=\"background-color:darkred\">darkred</span>, \n",
    "<span style=\"background-color:darksalmon\">darksalmon</span>, \n",
    "<span style=\"background-color:darkseagreen\">darkseagreen</span>, \n",
    "<span style=\"background-color:darkslateblue\">darkslateblue</span>, \n",
    "<span style=\"background-color:darkslategray\">darkslategray</span>, \n",
    "<span style=\"background-color:darkslategrey\">darkslategrey</span>, \n",
    "<span style=\"background-color:darkturquoise\">darkturquoise</span>, \n",
    "<span style=\"background-color:darkviolet\">darkviolet</span>, \n",
    "<span style=\"background-color:deeppink\">deeppink</span>, \n",
    "<span style=\"background-color:deepskyblue\">deepskyblue</span>, \n",
    "<span style=\"background-color:dimgray\">dimgray</span>, \n",
    "<span style=\"background-color:dimgrey\">dimgrey</span>, \n",
    "<span style=\"background-color:dodgerblue\">dodgerblue</span>, \n",
    "<span style=\"background-color:firebrick\">firebrick</span>, \n",
    "<span style=\"background-color:floralwhite\">floralwhite</span>, \n",
    "<span style=\"background-color:forestgreen\">forestgreen</span>, \n",
    "<span style=\"background-color:fuchsia\">fuchsia</span>, \n",
    "<span style=\"background-color:gainsboro\">gainsboro</span>, \n",
    "<span style=\"background-color:ghostwhite\">ghostwhite</span>, \n",
    "<span style=\"background-color:gold\">gold</span>, \n",
    "<span style=\"background-color:goldenrod\">goldenrod</span>, \n",
    "<span style=\"background-color:gray\">gray</span>, \n",
    "<span style=\"background-color:grey\">grey</span>, \n",
    "<span style=\"background-color:green\">green</span>, \n",
    "<span style=\"background-color:greenyellow\">greenyellow</span>, \n",
    "<span style=\"background-color:honeydew\">honeydew</span>,\n",
    "<span style=\"background-color:hotpink\">hotpink</span>, \n",
    "<span style=\"background-color:indianred\">indianred</span>, \n",
    "<span style=\"background-color:indigo\">indigo</span>, \n",
    "<span style=\"background-color:ivory\">ivory</span>, \n",
    "<span style=\"background-color:khaki\">khaki</span>, \n",
    "<span style=\"background-color:lavender\">lavender</span>, \n",
    "<span style=\"background-color:lavenderblush\">lavenderblush</span>, \n",
    "<span style=\"background-color:lawngreen\">lawngreen</span>, \n",
    "<span style=\"background-color:lemonchiffon\">lemonchiffon</span>, \n",
    "<span style=\"background-color:lightblue\">lightblue</span>, \n",
    "<span style=\"background-color:lightcoral\">lightcoral</span>, \n",
    "<span style=\"background-color:lightcyan\">lightcyan</span>, \n",
    "<span style=\"background-color:lightgoldenrodyellow\">lightgoldenrodyellow</span>, \n",
    "<span style=\"background-color:lightgray\">lightgray</span>, \n",
    "<span style=\"background-color:lightgrey\">lightgrey</span>, \n",
    "<span style=\"background-color:lightgreen\">lightgreen</span>, \n",
    "<span style=\"background-color:lightpink\">lightpink</span>, \n",
    "<span style=\"background-color:lightsalmon\">lightsalmon</span>, \n",
    "<span style=\"background-color:lightseagreen\">lightseagreen</span>, \n",
    "<span style=\"background-color:lightskyblue\">lightskyblue</span>, \n",
    "<span style=\"background-color:lightslategray\">lightslategray</span>, \n",
    "<span style=\"background-color:lightslategrey\">lightslategrey</span>, \n",
    "<span style=\"background-color:lightsteelblue\">lightsteelblue</span>, \n",
    "<span style=\"background-color:lightyellow\">lightyellow</span>, \n",
    "<span style=\"background-color:lime\">lime</span>, \n",
    "<span style=\"background-color:limegreen\">limegreen</span>, \n",
    "<span style=\"background-color:linen\">linen</span>, \n",
    "<span style=\"background-color:magenta\">magenta</span>, \n",
    "<span style=\"background-color:maroon\">maroon</span>, \n",
    "<span style=\"background-color:mediumaquamarine\">mediumaquamarine</span>, \n",
    "<span style=\"background-color:ediumblue\">ediumblue</span>, \n",
    "<span style=\"background-color:mediumorchid\">mediumorchid</span>, \n",
    "<span style=\"background-color:mediumpurple\">mediumpurple</span>,\n",
    "<span style=\"background-color:mediumseagreen\">mediumseagreen</span>, \n",
    "<span style=\"background-color:mediumslateblue\">mediumslateblue</span>, \n",
    "<span style=\"background-color:mediumspringgreen\">mediumspringgreen</span>, \n",
    "<span style=\"background-color:mediumturquoise\">mediumturquoise</span>, \n",
    "<span style=\"background-color:mediumvioletred\">mediumvioletred</span>, \n",
    "<span style=\"background-color:midnightblue\">midnightblue</span>,\n",
    "<span style=\"background-color:mintcream\">mintcream</span>, \n",
    "<span style=\"background-color:mistyrose\">mistyrose</span>, \n",
    "<span style=\"background-color:moccasin\">moccasin</span>, \n",
    "<span style=\"background-color:navajowhite\">navajowhite</span>, \n",
    "<span style=\"background-color:navy\">navy</span>, \n",
    "<span style=\"background-color:oldlace\">oldlace</span>, \n",
    "<span style=\"background-color:olive\">olive</span>, \n",
    "<span style=\"background-color:olivedrab\">olivedrab</span>, \n",
    "<span style=\"background-color:orange\">orange</span>, \n",
    "<span style=\"background-color:orangered\">orangered</span>, \n",
    "<span style=\"background-color:orchid\">orchid</span>, \n",
    "<span style=\"background-color:palegoldenrod\">palegoldenrod</span>, \n",
    "<span style=\"background-color:palegreen\">palegreen</span>, \n",
    "<span style=\"background-color:paleturquoise\">paleturquoise</span>, \n",
    "<span style=\"background-color:palevioletred\">palevioletred</span>, \n",
    "<span style=\"background-color:papayawhip\">papayawhip</span>, \n",
    "<span style=\"background-color:peachpuff\">peachpuff</span>, \n",
    "<span style=\"background-color:peru\">peru</span>, \n",
    "<span style=\"background-color:pink\">pink</span>, \n",
    "<span style=\"background-color:plum\">plum</span>, \n",
    "<span style=\"background-color:powderblue\">powderblue</span>, \n",
    "<span style=\"background-color:purple\">purple</span>, \n",
    "<span style=\"background-color:red\">red</span>, \n",
    "<span style=\"background-color:rosybrown\">rosybrown</span>, \n",
    "<span style=\"background-color:royalblue\">royalblue</span>, \n",
    "<span style=\"background-color:rebeccapurple\">rebeccapurple</span>, \n",
    "<span style=\"background-color:saddlebrown\">saddlebrown</span>, \n",
    "<span style=\"background-color:salmon\">salmon</span>, \n",
    "<span style=\"background-color:sandybrown\">sandybrown</span>, \n",
    "<span style=\"background-color:seagreen\">seagreen</span>, \n",
    "<span style=\"background-color:seashell\">seashell</span>, \n",
    "<span style=\"background-color:sienna\">sienna</span>, \n",
    "<span style=\"background-color:silver\">silver</span>, \n",
    "<span style=\"background-color:skyblue\">skyblue</span>, \n",
    "<span style=\"background-color:slateblue\">slateblue</span>, \n",
    "<span style=\"background-color:slategray\">slategray</span>, \n",
    "<span style=\"background-color:slategrey\">slategrey</span>, \n",
    "<span style=\"background-color:snow\">snow</span>, \n",
    "<span style=\"background-color:springgreen\">springgreen</span>, \n",
    "<span style=\"background-color:steelblue\">steelblue</span>, \n",
    "<span style=\"background-color:tan\">tan</span>, \n",
    "<span style=\"background-color:teal\">teal</span>, \n",
    "<span style=\"background-color:thistle\">thistle</span>, \n",
    "<span style=\"background-color:tomato\">tomato</span>, \n",
    "<span style=\"background-color:turquoise\">turquoise</span>, \n",
    "<span style=\"background-color:violet\">violet</span>, \n",
    "<span style=\"background-color:wheat\">wheat</span>, \n",
    "<span style=\"background-color:white\">white</span>, \n",
    "<span style=\"background-color:whitesmoke\">whitesmoke</span>, \n",
    "<span style=\"background-color:yellow\">yellow</span>, \n",
    "<span style=\"background-color:yellowgreen\">yellowgreen</span>\n",
    "\n",
    "\n",
    "**Built-in Plotly colorscales**  \n",
    "Select one of the following named colorscales for coloring the Heatmap section:\n",
    "\n",
    "'aggrnyl', 'agsunset', 'algae', 'amp', 'armyrose', 'balance',\n",
    "'blackbody', 'bluered', 'blues', 'blugrn', 'bluyl', 'brbg',\n",
    "'brwnyl', 'bugn', 'bupu', 'burg', 'burgyl', 'cividis', 'curl',\n",
    "'darkmint', 'deep', 'delta', 'dense', 'earth', 'edge', 'electric',\n",
    "'emrld', 'fall', 'geyser', 'gnbu', 'gray', 'greens', 'greys',\n",
    "'haline', 'hot', 'hsv', 'ice', 'icefire', 'inferno', 'jet',\n",
    "'magenta', 'magma', 'matter', 'mint', 'mrybm', 'mygbm', 'oranges',\n",
    "'orrd', 'oryel', 'oxy', 'peach', 'phase', 'picnic', 'pinkyl',\n",
    "'piyg', 'plasma', 'plotly3', 'portland', 'prgn', 'pubu', 'pubugn',\n",
    "'puor', 'purd', 'purp', 'purples', 'purpor', 'rainbow', 'rdbu',\n",
    "'rdgy', 'rdpu', 'rdylbu', 'rdylgn', 'redor', 'reds', 'solar',\n",
    "'spectral', 'speed', 'sunset', 'sunsetdark', 'teal', 'tealgrn',\n",
    "'tealrose', 'tempo', 'temps', 'thermal', 'tropic', 'turbid',\n",
    "'turbo', 'twilight', 'viridis', 'ylgn', 'ylgnbu', 'ylorbr', 'ylorrd'\n",
    "\n",
    "*Appending '_r' to a named colorscale reverses it.*\n",
    "\n",
    "You can also create your own colorscale as an array of lists, each composed of two elements. The first refers to the percentile rank, and the second to the applied color. Color can be provided in any string format among hex, hsl, rgb, or common name.\n",
    "<br>e.g.,\n",
    "<br>**custom_colorscale = [[0.0, '#636EFA'], [0.25, '#AB63FA'], [0.5, '#FFFFFF'], [0.75, '#E763FA'], [1.0, '#EF553B']]**\n",
    "\n",
    "\n",
    "Uncomment the following code cell and run it to display a preview for the built-in Plotly colorscales."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ad8a5426-2079-4d25-8469-e14c1b88e15d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import plotly.express as px\n",
    "#plotly_colors = px.colors.sequential.swatches_continuous()\n",
    "#plotly_colors.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "570fbbd2-ef97-4d4b-994c-d6a0f49a4f5f",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 0. Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "48aa3969-56ac-475a-ae79-652d5df1244a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import flask\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import json\n",
    "from urllib import request\n",
    "\n",
    "import dash\n",
    "from dash import html, dcc\n",
    "from dash.dependencies import Input, Output, State\n",
    "from dash.exceptions import PreventUpdate\n",
    "from dash.dash import no_update\n",
    "import dash_bootstrap_components as dbc\n",
    "import dash_bio as db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0d09cbd8-4ea6-4c3d-b0a5-6271a06dbaa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# input file\n",
    "#file_path = os.getcwd()       # [string] provide string path to the input file, os.getcwd() takes current path\n",
    "file_path = \"/Users/abadacz/REPOS/GIF/data_graphing/db_clustergram/data\"\n",
    "filename = \"input.csv\"        # [string] input file name\n",
    "col_separator = ','           # [string] type of the column separator in the input file: '\\t' for tabulator, ' ' for space, ',' for comma, etc.\n",
    "\n",
    "input_path = file_path\n",
    "df = pd.read_csv(str(input_path+\"/\"+filename).replace('//','/'), sep=col_separator)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cca3c391-9686-4142-ace7-3b5d65bb6133",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_html_label(text):\n",
    "    return html.Label(children=str(text), style=css_lab1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0be57331-2fbd-42c1-a0aa-07be818009f3",
   "metadata": {},
   "source": [
    "## 1. CSS styles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0060378d-1914-4de2-ace3-30f1fc3843e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "color1 = '#D6F2FA'    # light blue *\n",
    "color2 = '#738fab'    # middle teal\n",
    "color3 = '#008CBA'    # dark teal *\n",
    "color4 = '#006699'    # darkest teal\n",
    "color5 = ''\n",
    "css_div = {'display':'inline-block'}\n",
    "css_lpd = {**css_div, 'height':'95vh', 'padding':'5px', 'background-color':color1, 'border-right': '1px solid #008CBA'}\n",
    "css_rpd = {**css_div, 'flex-grow': '1', 'background-color':'ghostwhite'}\n",
    "css_btn = {'font-size':'20px', 'background-color':color3, 'color':'white', 'border-radius':'8px', 'border':'1px solid #006B88', 'marginBottom':'10px'}\n",
    "css_lab = {'font-style':'italic', 'display':'block', 'margin-top':'5px'}\n",
    "css_lab1 = {**css_lab, 'color':color3, 'font-size':'16px'}\n",
    "css_lab2 = {**css_lab, 'color':color2, 'font-size':'12px'}\n",
    "css_inp = {'marginBottom':'10px', 'width':'60%', 'display':'inline-block', 'font-size':'14px', 'padding':'6px 0'}\n",
    "#css_val = {'width':'30%', 'marginBottom':'20px'}\n",
    "#css_rad = {'padding': '1vh 2.5vw 0 0'}\n",
    "#css_col = {'display':'inline-block', 'width':'8%', 'padding':'3px 3px 3px 3px', 'font-size':'12px'}\n",
    "\n",
    "css_upload_box = {'background-color': '#f9f9f9', 'width': '25vw', 'min-height': '60px', 'height': '20vh', 'border-width': '2px', 'border-color':color4, 'border-style': 'dashed', 'border-radius': '5px', 'display': 'table-cell', 'vertical-align': 'middle', 'text-align': 'center'}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "962f1e8e-d2cc-47c1-93cc-5bb11be8a9f0",
   "metadata": {},
   "source": [
    "## 2. App layout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "17ab3743-ca73-4c62-a3ca-649a40e5963d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Layout components\n",
    "opts1 = [\n",
    "  html.Label('from file system: ', id='custom-label', style=css_lab2),\n",
    "  dcc.Upload(id='upload-box',\n",
    "    children=html.Div([html.P('drag & drop', style={'color':color2}), html.P('or click to browse', style={'color':color2})]),\n",
    "    className='upload-box',\n",
    "    style=css_upload_box,\n",
    "    max_size=-1,\n",
    "    multiple=True),\n",
    "  html.Label('from online resources: ', id='custom-url-label', style=css_lab2),\n",
    "  dcc.Input(id=\"custom-url\", type=\"url\", placeholder=\"enter URL\", value='', debounce=False),\n",
    "  dbc.Button(\"download\", id=\"download-btn\", n_clicks=0, size=\"sm\", outline=True, color=\"secondary\", className=\"me-1 align-top\", style={'marginLeft':'1vw', 'width':'25%'}),\n",
    "  html.P(''),\n",
    "  html.Label(id=\"settings-upload-label\", style=css_lab1),\n",
    "  html.Div(id=\"settings-upload-inputs\", className=\"mt-3\")\n",
    "]\n",
    "\n",
    "opts = html.Div([\n",
    "  dbc.Accordion([\n",
    "    dbc.AccordionItem(\n",
    "      opts1,\n",
    "      title=\"A. UPLOAD INPUTS\",\n",
    "      item_id=\"item-1\",\n",
    "    ),\n",
    "    dbc.AccordionItem(\n",
    "      \"This is the content of the second section\",\n",
    "      title=\"Item 2\",\n",
    "      item_id=\"item-2\",\n",
    "    ),\n",
    "    dbc.AccordionItem(\n",
    "      \"This is the content of the third section\",\n",
    "      title=\"Item 3\",\n",
    "      item_id=\"item-3\",\n",
    "    ),\n",
    "  ], id=\"accordion\", start_collapsed=True, always_open=True, flush=False, style={'width':'25vw'}),\n",
    "  html.Div(id=\"accordion-contents\", className=\"mt-3\"),\n",
    "], id='optionsDiv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1aa2a210-a22b-44e0-b2da-06981c5d55bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "app = dash.Dash(__name__, external_stylesheets=[dbc.themes.BOOTSTRAP])\n",
    "\n",
    "app.scripts.config.serve_locally=True\n",
    "app.css.config.serve_locally=True\n",
    "\n",
    "app.css.append_css({'external_url':'/Users/abadacz/REPOS/GIF/data_graphing/_assets/css/stylesheet.css'})\n",
    "\n",
    "#@app.server.route('/static/<path:path>')\n",
    "#def static_file(path):\n",
    "#    static_folder = os.path.join(os.getcwd(), 'static')\n",
    "#    return send_from_directory(static_folder, path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e4783ea7-2a2d-4fd5-b70f-1b68e9be0633",
   "metadata": {},
   "outputs": [],
   "source": [
    "app.layout = html.Div([\n",
    "  html.Link(href='stylesheet.css', rel='stylesheet'),\n",
    "  # STORAGE\n",
    "  dcc.Store(id=\"custom-box-files\", data='', storage_type='session'),\n",
    "  dcc.Store(id=\"custom-url-files\", data='', storage_type='session'),\n",
    "  dcc.Store(id=\"user-files-list\", data='', storage_type='session'),\n",
    "\n",
    "  # OPTIONS\n",
    "  html.Div([\n",
    "    html.Button('Show Options', id='options', n_clicks=0, style=css_btn),\n",
    "    opts,    \n",
    "  ], id='left-panelDiv', style=css_lpd),\n",
    "\n",
    "  # GRAPH\n",
    "  html.Div([\n",
    "#    db.Clustergram(data=df),\n",
    "#    db.Ideogram(id='dashbio-ideogram',),\n",
    "  ], id='graph-panelDiv', style=css_rpd)\n",
    "\n",
    "], id='app', style={'display': 'flex'})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e54974b7-6e3b-43c0-9382-84c6d5f221f5",
   "metadata": {},
   "source": [
    "## 3. Javascript clientside callbacks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c59fc5fd-dbbb-4d7f-9363-f82035c22ef5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# open-close options section\n",
    "app.clientside_callback(\n",
    "    \"\"\"\n",
    "    function(largeValue1, largeValue2) {\n",
    "        var x = document.getElementById(\"optionsDiv\");\n",
    "        var y = document.getElementById(\"options\");\n",
    "        var z = document.getElementById(\"left-panelDiv\");\n",
    "        if (x.style.display === \"none\") {\n",
    "            x.style.display = \"block\";\n",
    "            y.style.backgroundColor = \"#D6F2FA\";\n",
    "            y.style.color = \"#90B6C1\";\n",
    "            y.innerText = \"âœ•\";\n",
    "            z.width = \"25vw\";\n",
    "        } else {\n",
    "            x.style.display = \"none\";\n",
    "            y.style.backgroundColor = \"#008CBA\";\n",
    "            y.style.color = \"white\";\n",
    "            y.innerText = \"â‰¡\";\n",
    "            z.width = \"2.1%\";\n",
    "        }\n",
    "    }\n",
    "    \"\"\",\n",
    "    Output('optionsDiv', 'style'),\n",
    "    Input('options', 'n_clicks'),\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d679ea70-a1bb-453e-9318-181ce8e69f94",
   "metadata": {},
   "source": [
    "## 4. Callbacks responsive to changes in Dash widgets (options panel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9bfdba4d-e192-44fe-a0cf-32b722032b96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n",
    "@app.callback(\n",
    "    Output(\"accordion-contents\", \"children\"),\n",
    "    [Input(\"accordion-options\", \"active_item\")],\n",
    ")\n",
    "def change_item(item):\n",
    "    return f\"Item selected: {item}\"\n",
    "\n",
    "# The function returns updated dictionary of filename:file_content for all user-loaded files via upload-box\n",
    "@app.callback(Output('custom-box-files', 'data'),\n",
    "              [Input('upload-box', 'filename'), Input('upload-box', 'contents')],\n",
    "              State('custom-box-files', 'data'), prevent_initial_call = True)\n",
    "def upload_files_box(names, contents, files):\n",
    "    if names is not None:\n",
    "        if files == '':\n",
    "            files = {i:contents[num] for num,i in enumerate(names)}\n",
    "        else:\n",
    "            files = json.loads(files)\n",
    "            for num,i in enumerate(names):\n",
    "                files[i] = contents[num]\n",
    "        return json.dumps(files)\n",
    "\n",
    "# The function returns updated dictionary of filename:file_content for all user-loaded files via online-url\n",
    "@app.callback(Output('custom-url-files', 'data'),\n",
    "              [Input('download-btn', 'n_clicks')],\n",
    "              [State('custom-url', 'value'), State('custom-url-files', 'data')], prevent_initial_call = True)\n",
    "def upload_files_url(download, url, files):\n",
    "    if url != '':\n",
    "        filename = url.strip().split('/')[-1]\n",
    "        try:\n",
    "            with request.urlopen(url) as f:\n",
    "                content = f.read().decode('utf-8')\n",
    "            if files == '':\n",
    "                files = {}\n",
    "                files[filename] = content\n",
    "            else:\n",
    "                files = json.loads(files)\n",
    "                files[filename] = content\n",
    "            return json.dumps(files)\n",
    "        except:\n",
    "            pass\n",
    "    else:\n",
    "        raise PreventUpdate\n",
    "\n",
    "# The function creates a list of user-loaded files that can be used for plotting\n",
    "@app.callback(Output('user-files-list', 'data'),\n",
    "             [Input('custom-box-files', 'data'), Input('custom-url-files', 'data')], prevent_initial_call = True)\n",
    "def create_list_of_custom_files(files_box, files_url):\n",
    "    files = []\n",
    "    for i in [files_box, files_url]:\n",
    "        if i != '':\n",
    "            files.extend(list(json.loads(i).keys()))\n",
    "    return files\n",
    "\n",
    "# Display summary of settings\n",
    "@app.callback(\n",
    "    [Output(\"settings-upload-label\", \"children\"), Output(\"settings-upload-inputs\", \"children\")],\n",
    "    [Input(\"user-files-list\", \"data\")])\n",
    "def display_inputs_settings(loaded_files):\n",
    "    if loaded_files == '':\n",
    "        return[\"Please load inputs using available options.\", '']\n",
    "    else:\n",
    "        info = \"The following files were loaded: \"\n",
    "        return [info, [generate_html_label(i) for i in loaded_files]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ff3906c-bf68-4cb3-9a09-42d1f1f3f24d",
   "metadata": {},
   "source": [
    "## 5. Run server"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a081c28-9469-4dff-926d-3ce4d777052f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dash is running on http://127.0.0.1:8050/\n",
      "\n",
      " * Serving Flask app '__main__' (lazy loading)\n",
      " * Environment: production\n",
      "\u001b[31m   WARNING: This is a development server. Do not use it in a production deployment.\u001b[0m\n",
      "\u001b[2m   Use a production WSGI server instead.\u001b[0m\n",
      " * Debug mode: off\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " * Running on http://127.0.0.1:8050/ (Press CTRL+C to quit)\n",
      "/Users/abadacz/Library/Miniforge3_x86/envs/graphing/lib/python3.9/site-packages/dash/resources.py:63: UserWarning: You have set your config to `serve_locally=True` but A local version of /Users/abadacz/REPOS/GIF/data_graphing/_assets/css/stylesheet.css is not available.\n",
      "If you added this file with `app.scripts.append_script` or `app.css.append_css`, use `external_scripts` or `external_stylesheets` instead.\n",
      "See https://dash.plotly.com/external-resources\n",
      "  warnings.warn(\n",
      "127.0.0.1 - - [13/Dec/2022 17:36:17] \"GET / HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Dec/2022 17:36:17] \"\u001b[36mGET /_dash-component-suites/dash/dcc/async-upload.js HTTP/1.1\u001b[0m\" 304 -\n",
      "/Users/abadacz/Library/Miniforge3_x86/envs/graphing/lib/python3.9/site-packages/dash/resources.py:63: UserWarning:\n",
      "\n",
      "You have set your config to `serve_locally=True` but A local version of /Users/abadacz/REPOS/GIF/data_graphing/_assets/css/stylesheet.css is not available.\n",
      "If you added this file with `app.scripts.append_script` or `app.css.append_css`, use `external_scripts` or `external_stylesheets` instead.\n",
      "See https://dash.plotly.com/external-resources\n",
      "\n",
      "127.0.0.1 - - [13/Dec/2022 17:36:17] \"GET /stylesheet.css HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Dec/2022 17:36:17] \"GET /_dash-layout HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Dec/2022 17:36:17] \"GET /_dash-dependencies HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Dec/2022 17:36:17] \"POST /_dash-update-component HTTP/1.1\" 200 -\n"
     ]
    }
   ],
   "source": [
    "app.run_server(debug=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8555a5a9-952a-4781-9cbd-8c869abc58c9",
   "metadata": {},
   "source": [
    "## 1. Set up global variables\n",
    "You can easily adjust the overall styling of the chart by setting up the global variables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38c4a1e2-a07d-4a96-8ee4-816fcac65d86",
   "metadata": {},
   "outputs": [],
   "source": [
    "# input file\n",
    "file_path = os.getcwd()       # [string] provide string path to the input file, os.getcwd() takes current path\n",
    "filename = \"input.csv\"        # [string] input file name\n",
    "col_separator = ','           # [string] type of the column separator in the input file: '\\t' for tabulator, ' ' for space, ',' for comma, etc.\n",
    "class_labs = []               # [list] (optional) list of labels for additional category; should match the number of data columns in the input file;\n",
    "\n",
    "# initial data parsing and clustering (to generate data for dendrograms)\n",
    "log_transform = False         # [boolean] if True, ransform the data by taking the base-two logarithm of all values in the dataset\n",
    "standardize = 'none'          # [string] default = 'none'; dimension for standardizing values: 'none', 'row', 'column'; with standardization, the mean is 0 and the standard deviation is 1;\n",
    "cluster = 'all'               # [string] default = 'all'; dimension along which the data will be clustered: 'row', 'column', or 'all';\n",
    "row_dist = 'euclidean'        # [string] default = 'euclidean'; distance metric for rows: braycurtisâ€™, â€˜canberraâ€™, â€˜chebyshevâ€™, â€˜cityblockâ€™, â€˜correlationâ€™, â€˜cosineâ€™, â€˜diceâ€™, â€˜euclideanâ€™, â€˜hammingâ€™, â€˜jaccardâ€™, â€˜jensenshannonâ€™, â€˜kulsinskiâ€™, â€˜kulczynski1â€™, â€˜mahalanobisâ€™, â€˜matchingâ€™, â€˜minkowskiâ€™, â€˜rogerstanimotoâ€™, â€˜russellraoâ€™, â€˜seuclideanâ€™, â€˜sokalmichenerâ€™, â€˜sokalsneathâ€™, â€˜sqeuclideanâ€™, â€˜yuleâ€™ (works when using scipy.spatial.distance.pdist()) \n",
    "col_dist = 'euclidean'        # [string] default = 'euclidean'; distance metric for columns: braycurtisâ€™, â€˜canberraâ€™, â€˜chebyshevâ€™, â€˜cityblockâ€™, â€˜correlationâ€™, â€˜cosineâ€™, â€˜diceâ€™, â€˜euclideanâ€™, â€˜hammingâ€™, â€˜jaccardâ€™, â€˜jensenshannonâ€™, â€˜kulsinskiâ€™, â€˜kulczynski1â€™, â€˜mahalanobisâ€™, â€˜matchingâ€™, â€˜minkowskiâ€™, â€˜rogerstanimotoâ€™, â€˜russellraoâ€™, â€˜seuclideanâ€™, â€˜sokalmichenerâ€™, â€˜sokalsneathâ€™, â€˜sqeuclideanâ€™, â€˜yuleâ€™ (works when using scipy.spatial.distance.pdist()) \n",
    "link_method = 'complete'      # [string] default = 'complete'; linkage algorithm to use: 'complete', 'average', 'weighted' and 'ward' (nearest-neighbors chain), or 'single' (optimized minimum spanning)\n",
    "\n",
    "######- interactive graph : DENDROGRAM section(s) -######\n",
    "leaf_order = True             # [boolean] determine leaf order that maximizes similarity between neighboring leaves\n",
    "dendro_ratio = [0.15, 0.3]    # [list] list of two float elements providing proportions of heatmap vs. dendrogram sections; scale ratio of [X axis dendrogram, Y axis dendrogram] with respect to the size of the heatmap\n",
    "dendro_factor = 0.9           # [float] scale ratio between dendrogram and cluster bars sections\n",
    "dendro_line_width = [2, 2]    # [list] line width [row, column] for the dendrograms\n",
    "dendro_threshold = {'row': 3, 'col': 3}   # maximum linkage value for which unique colors are assigned to clusters\n",
    "dendro_colors = {'row': ['#636EFA', '#00CC96', '#19D3F3'], 'col': ['#AB63FA', '#EF553B'], 'bg': '#506784'}      # list of colors to use for different clusters in the dendrogram that have a root under the threshold for each dimension \n",
    "row_prefix = 'protein cluster' # custom name of the row dendrogram clusters\n",
    "col_prefix = 'sample cluster'  # custom name of the column dendrogram clusters\n",
    "list_all_row = True           # [boolean] on interactive labels for Y dendrogram list all elements in the given cluster\n",
    "n_per_line_row = 5            # [int] number of elements names listed per single line on interactive labels for Y dendrogram (prevents overly wide labels )\n",
    "\n",
    "######- interactive graph : GROUP/CLASS BARS -###### \n",
    "row_markers = [{'group': 98, 'annotation': 'description for cluster 98', 'color': 'white'}, {'group': 75, 'annotation': 'description for cluster 75', 'color': '#009dd1'}, {'group': 1, 'annotation': 'description for cluster 1', 'color': 'white'}]    # bars for selected row (Y) clusters; 'white' for 'color' will use by default the corresponding color from dendrogram; use <br> tag within annotation to break the string into multiple lines; \n",
    "col_markers = [{'group': 1, 'annotation': 'description for cluster 1', 'color': 'white'}, {'group': 2, 'annotation': 'description for cluster 2', 'color': 'yellowgreen'}, {'group': 6, 'annotation': 'description for cluster 6', 'color': 'white'}]       # bars for selected column (X) clusters; 'white' for 'color' will use by default the corresponding color from dendrogram; use <br> tag within annotation to break the string into multiple lines; \n",
    "\n",
    "\n",
    "######- interactive graph : HEATMAP SECTION -######\n",
    "# - HEATMAP GENERAL SETTINGS\n",
    "center_values = True          # [boolean] if True, center the values of the heatmap about zero\n",
    "display_cutoff = 2            # [int] standardized values below the negative of this value will be colored with one shade, and the values that are above this value will be colored with another\n",
    "color_map = 'rdbu'            # colorscale for the heatmap; use built-in Plotly colorscale (list provided on the top of this notebook) or define your own colorscale\n",
    "\n",
    "# - HEATMAP COLORBAR SETTINGS\n",
    "show_colorbar = True                     # [boolean] if True, the colorbar (heatmap colorscale) will be displayed;\n",
    "colorbar_len = 1 - dendro_ratio[0]       # [float] vertical length of the colorbar; default = 1 - dendro_ratio[0], means that the colorbar height matches the heatmap height\n",
    "colorbar_x = 1.04                        # [float] horizontal position of the colorbar; value above 1 places colorbar on the right margin of the plotting area\n",
    "colorbar_y = 0.5*(1 - dendro_ratio[0])   # [float] vertical position of the center of the colorbar; default = 0.5*(1-dendro_ratio[0]), means that the colorbar will be aligned vertically with the heatmap \n",
    "\n",
    "# - HEATMAP X AXIS SETTINGS\n",
    "#.... X axis title\n",
    "x_title = \"X: Sample identifier\"         # [string] text of X axis title\n",
    "x_ti_color = \"black\"                     # [string] color of X axis title\n",
    "x_ti_size = 20                           # [int] color of X axis title\n",
    "\n",
    "#.... X axis line (graph border)\n",
    "x_axis_show = True            # [boolean] if True the X axis is visible, if False X axis is hidden\n",
    "x_axis_mirror = False         # [boolean] if True the X axis complementary mirror axis is visible (on the top/bottom of the graph)\n",
    "x_axis_width = 1              # [int] width of X axis line\n",
    "x_axis_color = \"black\"        # [string] color of X axis line\n",
    "\n",
    "#.... X axis ticks' labels\n",
    "show_xtick_lab = True         # [string] if True the labels of X axis ticks are visible\n",
    "xtick_labs = []               # [list] list of strings with ticks' labels for X axis; should match the number of data columns in the input file; if empty list, the keys (columns' names) from the header will be used\n",
    "x_tick_lab_side = 'bottom'    # [string] side of the graph where the ticks and labels will apear; enter \"top\" or 'bottom'\n",
    "x_tick_lab_angle = 45         # [int] angle of X axis ticks in range 0 - 360\n",
    "x_tick_lab_size = 14          # [int] font size of ticks labels for X axis (size must correspond to the list of tickvals) \n",
    "x_tick_lab_color = \"black\"    # [string] font color of ticks labels for X axis \n",
    "\n",
    "#.... X axis ticks (marks)\n",
    "x_axis_ticks = \"outside\"      # [string] position of X axis ticks: \"outside\", \"inside\", \"\" (not visible)\n",
    "x_tick_width = 1              # [int] width (in px) of X axis ticks\n",
    "x_tick_len = 5                # [int] length (in px) of X axis ticks\n",
    "x_tick_color = 'black'        # [string] color of X axis ticks\n",
    "\n",
    "# - HEATMAP Y AXIS SETTINGS\n",
    "#.... Y axis title\n",
    "y_title = \"Y: Protein UniProt ID\"      # [string] custom title for Y axis\n",
    "y_ti_color = \"black\"                   # [string] color for Y axis title\n",
    "y_ti_size = 20                         # [int] color for Y axis title\n",
    "y_ti_x = 1.24                          # [float] Y axis title horizontal position; balow 0 - left side of the graph, above 1 - right side of the graph\n",
    "y_ti_y = 0.42                          # [float] Y axis title vertical position; usually value around 0.5 center title with the heatmap\n",
    "y_ti_angle = 90                        # [float] the angle of the text: 0 is horizontal; -90 is vertical-left; 90 is vertical-right\n",
    "\n",
    "#.... Y axis line (graph border)\n",
    "y_axis_show = True            # [boolean] if True the Y axis is visible, if False Y axis is hidden\n",
    "y_axis_mirror = False         # boolean] if True the X axis complementary mirror axis is visible (on the left/right of the graph)\n",
    "y_axis_width = 1              # [int] width of Y axis line\n",
    "y_axis_color = \"black\"        # [string] color of Y axis line\n",
    "\n",
    "#.... Y axis ticks' labels\n",
    "show_ytick_lab = True         # [string] if True the labels of Y axis ticks are visible\n",
    "y_tick_lab_side = 'right'     # [string] side of the graph where the ticks and labels will apear; enter \"left\" or 'right'\n",
    "y_tick_lab_angle = 0          # [int] angle of Y axis ticks in range 0 - 360\n",
    "y_tick_lab_size = 12          # [int] font size of ticks labels for Y axis (size must correspond to the list of tickvals) \n",
    "y_tick_lab_color = \"black\"    # [string] font color of ticks labels for Y axis \n",
    "\n",
    "y_ticks_str = True            # [boolean] if True, text labels for the Y axis in the heatmap section can be dispalyed; otherwise an automatic numerical (data index) values will be dispalyed;\n",
    "y_col_name = 'label'          # [string] name of the column in the input used for labeling ticks on the Y-axis for the heatmap section; requires: y_ticks_str = True\n",
    "y_nticks = 10                 # [int] number of ticks on the zoomed Y axis with text ticks' labels; requires: y_ticks_str = True\n",
    "y_zoom = [0,1]                # [range] specify the zoomed fragment along normalized Y axis; 0 - bottom, 1 - top; requires: y_ticks_str = True; text labels switches off the interactive Y axis zooming - this variable can zoom the Y axis though; \n",
    "\n",
    "#.... Y axis ticks (marks)\n",
    "y_axis_ticks = \"outside\"      # [string] position of Y axis ticks: \"outside\", \"inside\", \"\" (not visible)\n",
    "y_tick_width = 1              # [int] width (in px) of Y axis ticks\n",
    "y_tick_len = 5                # [int] length (in px) of Y axis ticks\n",
    "y_tick_color = 'black'        # [string] color of Y axis ticks\n",
    "\n",
    "\n",
    "# - HEATMAP ON-HOVER INTERACTIVE LABELS\n",
    "heatmap_name = 'heatmap'      # [string] custom name displayed on the interactive labels for the heatmap points\n",
    "h_label_x = 'sample'          # [string] label for X data displayed on the interactive label; by default the ticks of X axis will be used as data; 'none' will switch off the option \n",
    "h_label_y = 'UniProt ID'      # [string] label for Y data displayed on the interactive label; by default the ticks of Y axis will be used as data; 'none' will switch off the option \n",
    "h_label_z = 'value'           # [string] label for Z data displayed on the interactive label; by default the heatmap (Z) values will be used as data; 'none' will switch off the option \n",
    "h_label_4 = 'none'            # ['none' or list] optional 4-element list [label, dimension, data, file]; label element is any string; dimension is a string: 'x', 'y', 'z' specyfing the data structure; data is an array (1D for 'x' or 'y', 2D for 'z') provided explicitly or as a string of the column name from the input file; provide file name if other than 'filename' variable\n",
    "h_label_5 = 'none'            # ['none' or list] optional 4-element list [label, dimension, data, file]; label element is any string; dimension is a string: 'x', 'y', 'z' specyfing the data structure; data is an array (1D for 'x' or 'y', 2D for 'z') provided explicitly or as a string of the column name from the input file; provide file name if other than 'filename' variable\n",
    "#ex# h_label_4 = ['x index', 'x', [str(i+1) for i in range(10)], '']    # example usage: add numerical index for samples on X axis (X dimension); use range(n) where n is a number of columns on your heatmap\n",
    "#ex# h_label_4 = ['custom y', 'y', y_col_name, '']                      # example usage: add lables stored in the selected column (Y dimension) in the input file; here the column name is provided via pre-defined variable\n",
    "#ex# h_label_4 = ['unique z', 'z', '', 'external_file.csv']             # example usage: add unique description for each XY point; use data stored in the other file of dimensions corresponding to the Heatmap input file (same number of columns and rows)\n",
    "\n",
    "\n",
    "######- EXPORT SETTINGS -######\n",
    "\n",
    "# interactive graph - on-hover labels\n",
    "hover_align = \"auto\"          # [string] horizontal alignment of the text content within hover label box: \"left\", \"right\", \"auto\"\n",
    "hover_size = 14               # [int] font size of interactive labels\n",
    "\n",
    "# interactive graph - general layout settings\n",
    "graph_height = 600            # [int] height of an interactive graph\n",
    "graph_width = 800             # [int] width of an interactive graph\n",
    "margins = [0,'d',350,'d']       # [list] int values for graph margins in order [left, top, right, bottom]; default = 'd'\n",
    "graph_bg_color = \"white\"      # [string] background color in the plottiong area e.g., common color name, rgb(r,g,b), \"#FFFFFF\")\n",
    "drawing_bg_color = \"white\"    # [string] background color of the drawing (e.g., common color name, rgb(r,g,b), \"#FFFFFF\")\n",
    "html_auto = False             # [boolean] if True the generated interactive HTML graph will automatically open in a separate tab in the web browser\n",
    "\n",
    "# interactive graph - title & legend settings\n",
    "graph_title = \"Clustergram Plot\"        # [string] provide custom title for Graph title\n",
    "font_family = \"Arial\"                   # [string] font family: \"Arial\", \"Balto\", \"Courier New\", \"Droid Sans\",, \"Droid Serif\", \"Droid Sans Mono\", \"Gravitas One\", \"Old Standard TT\", \"Open Sans\", \"Overpass\", \"PT Sans Narrow\", \"Raleway\", \"Times New Roman\"\n",
    "ti_color = \"black\"                      # [string] graph title color (e.g., common color name: \"blue\", rgb: rgb(0,0,0), hex: \"#FFFFFF\")\n",
    "ti_size = 24                            # [int] graph title font size\n",
    "ti_x = 0.38                             # [float] graph title horizontal position\n",
    "ti_y = 0.9                              # [float] graph title vertical position\n",
    "is_legend = True                        # [boolean] if True the legend is visible, if False the legend is hidden\n",
    "legend_x = 1.4\n",
    "legend_y = 1\n",
    "\n",
    "# static image export settings\n",
    "static_img_format = 'png'                      # [string] format of exported static image: png, svg, jpeg, webp, pdf, eps\n",
    "static_img_filename = 'example_clustergram'    # [string] provide custom filename for exported static image\n",
    "static_img_height = 1200                       # [int] enter the expected height (in px) of exported static image \n",
    "static_img_width = 800                         # [int] enter the expected width (in px) of exported static image \n",
    "static_img_scale = 1                           # [int] enter the expected scale ratio of exported static image \n",
    "\n",
    "print(filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3b226e4-ecdd-4263-9d66-db7c81bdcc84",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 2. Load input file(s) \n",
    "Read and load data into object-like DataFrame data structure with *Pandas*\n",
    "* input file: *input.csv* &emsp; &emsp; # the input file in this example is \",\" separated\n",
    "* input path: *`input_path=os.getcwd()`* &emsp; &emsp; # finds the path to the current directory <br>\n",
    "    *<span style=\"color:gray\">If your input is in the children-directory of the current folder you can easily append the remaining part of the path: <br>`path=os.getcwd()+'/INPUTS'`, and if it is on a completely different path, then give the full path.</span>* \n",
    "* dataframe: *`df=pd.DataFrame()`* &emsp; &emsp; # object like data structure created with pandas library\n",
    "    * *`pd.read_csv(filename, sep=\",\")`* &emsp; # reads column-type text file, default separator is comma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b24c7c0-8317-4d27-af09-a6c4e2aefcf7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "input_path = file_path\n",
    "df = pd.read_csv(str(input_path+\"/\"+filename).replace('//','/'), sep=col_separator)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d2932a7-13e0-4feb-9421-9d6610c6e5b2",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 2a.(optional) Preview your data structure\n",
    "\n",
    "* *`print(df)`* &emsp; # displays the top...bottom fragments of pandas *DataFrame* object\n",
    "* *`print(df.keys())`* &emsp; # displays the names of the columns<br>\n",
    "    <span style=\"color:gray\">*DataFrame* keys are very important, because they are directly used to select the column(s) from which the data will be extracted to the chart.</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93989eab-81bd-4a07-a856-067b70e20e3a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "print(\"The keys in your DataFrame are: \\n\\n\", df.keys().tolist(), \"\\n\")\n",
    "#print(\"There is a preview of your DataFrame: \\n\\n\", df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07ddba7c-c49e-41e9-8ae6-f3109bda37c4",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 3. (optional) Parse header row in the dataframe to adjust keys (column names)\n",
    "You can modyfiy the header row in the pandas dataframe to adjust column names (used as labels on X axis in the heatmap section) or create new useful variables. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8ba4c7d-ab68-4e06-9154-f2fe90c06448",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "if not xtick_labs:\n",
    "    labels = list(df.columns.values)         # derive heder row\n",
    "    labels.remove(y_col_name)                # remove key of the data labeling column\n",
    "    xtick_labs = ['_'.join(lab.split('_')[:3]) for lab in labels]              # modify each key keeping relevant part of the string (primary classification category)\n",
    "    if not class_labs:\n",
    "        class_labs = [lab.split('_')[3].replace('.','_') for lab in labels]    # modify each key keeping the other part of the string (secondary classification category - optional)\n",
    "    if h_label_5 == 'none':\n",
    "        h_label_5 = ['sample class', 'x', class_labs, '']                      # use extracted data (secondary classification) on the interactive labels of the Heatmap section\n",
    "#print(\"xtick_labs = \",xtick_labs)\n",
    "#print(\"class_labs = \",class_labs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "682ebb60-9ec4-4bac-981b-082ea19fdfcd",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 4. Create trace(s) for your graph \n",
    "In Plotly **trace** is a data series plotted as a particular type of graph. In this example we will use **clustergram** but there are many more predefined [trace types](https://plotly.com/python/reference/scatter/) in *Plotly*. <br>\n",
    "*<span style=\"color:gray\">In general, complex graphs composed of many traces are more convenient to create using Graph Objects wrapper. You can find a suitable example in this repository.</span>*<br>\n",
    "In this example, we will create an interactive clustergram composed of many traces, including single 2D heatmap, two 1D dendrograms for X & Y axes, and bar charts for additional level of data classification."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8094ac71",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "fig, curves, traces = db.Clustergram(   # define Figure object composed of a heatmap with dendogram\n",
    "    data=df,                            # provide source dataFrame\n",
    "    column_labels=xtick_labs,           # list of labels for the heatmap X axis\n",
    "    row_labels=list(df.index),          # list of numerical indices for the heatmap Y axis \n",
    "    generate_curves_dict = True,        # return a dictionary with the cluster number associated with each curve number in the graph; May be useful for capturing the cluster number that is clicked.\n",
    "    return_computed_traces = True,      # return the precomputed dendrogram traces; May be useful if one wishes to add, e.g., group markers to the figure without recalculating the clustering in the entire figure.\n",
    "\n",
    "    # data parsing and clustering\n",
    "    log_transform = log_transform,      # if True, ransform the data by taking the base-two logarithm of all values in the dataset\n",
    "    standardize = standardize,          # dimension for standardizing values: 'none', 'row', 'column'; with standardization, the mean is 0 and the standard deviation is 1;\n",
    "    cluster = cluster,                  # dimension along which the data will be clustered: 'row', 'column', or 'all';\n",
    "    row_dist = row_dist,                # distance metric for rows: braycurtisâ€™, â€˜canberraâ€™, â€˜chebyshevâ€™, â€˜cityblockâ€™, â€˜correlationâ€™, â€˜cosineâ€™, â€˜diceâ€™, â€˜euclideanâ€™, â€˜hammingâ€™, â€˜jaccardâ€™, â€˜jensenshannonâ€™, â€˜kulsinskiâ€™, â€˜kulczynski1â€™, â€˜mahalanobisâ€™, â€˜matchingâ€™, â€˜minkowskiâ€™, â€˜rogerstanimotoâ€™, â€˜russellraoâ€™, â€˜seuclideanâ€™, â€˜sokalmichenerâ€™, â€˜sokalsneathâ€™, â€˜sqeuclideanâ€™, â€˜yuleâ€™ (scipy.spatial.distance.pdist()) \n",
    "    col_dist = col_dist,                # distance metric for columns: braycurtisâ€™, â€˜canberraâ€™, â€˜chebyshevâ€™, â€˜cityblockâ€™, â€˜correlationâ€™, â€˜cosineâ€™, â€˜diceâ€™, â€˜euclideanâ€™, â€˜hammingâ€™, â€˜jaccardâ€™, â€˜jensenshannonâ€™, â€˜kulsinskiâ€™, â€˜kulczynski1â€™, â€˜mahalanobisâ€™, â€˜matchingâ€™, â€˜minkowskiâ€™, â€˜rogerstanimotoâ€™, â€˜russellraoâ€™, â€˜seuclideanâ€™, â€˜sokalmichenerâ€™, â€˜sokalsneathâ€™, â€˜sqeuclideanâ€™, â€˜yuleâ€™ (scipy.spatial.distance.pdist()) \n",
    "    link_method = link_method,          # linkage algorithm to use: 'complete', 'average', 'weighted' and 'ward' (nearest-neighbors chain), or 'single' (optimized minimum spanning)\n",
    "#    dist_fun = scipy.spatial.distance.pdist,      # loaded default = scipy.spatial.distance.pdist (Note: when uncommented the import of scipy is needed); function to compute the pairwise distance from the observations\n",
    "#    link_fun = scipy.cluster.hierarchy.linkage,   # loaded default = scipy.cluster.hierarchy.linkage (Note: when uncommented the import of scipy is needed); function to compute the linkage matrix from the pairwise distances\n",
    "\n",
    "    # dendrograms\n",
    "    optimal_leaf_order = leaf_order,    # determine leaf order that maximizes similarity between neighboring leaves\n",
    "    display_ratio = dendro_ratio,       # dendrograms' heights [row, column] with respect to the size of the heatmap\n",
    "    color_threshold = dendro_threshold, # maximum linkage value for which unique colors are assigned to clusters\n",
    "    color_list = dendro_colors,         # list of colors to use for different clusters in the dendrogram that have a root under the threshold for each dimension \n",
    "    line_width = dendro_line_width,     # line width [row, column] for the dendrograms\n",
    "\n",
    "    # classification bars\n",
    "    row_group_marker = row_markers,     # annotations for row clusters \n",
    "    col_group_marker = col_markers,     # annotations for column clusters \n",
    "    annotation_font = {},               # font options for annotations\n",
    "\n",
    "    # heatmap\n",
    "    center_values = center_values,      # if True, center the values of the heatmap about zero\n",
    "    display_range = display_cutoff,     # standardized values below the negative of this value will be colored with one shade, and the values that are above this value will be colored with another\n",
    "    color_map = color_map,              # colorscale for the heatmap\n",
    "    \n",
    "    # other optional params\n",
    "#    imputer_parameters = {'missing_values' : '', 'strategy' : '', 'axis' : ''}    # if needed, see https://scikit -learn.org/stable/modules/generated/sklearn.preprocessing.Imputer.html; 'axis=0' for columns, 'axis=1' for rows    \n",
    "#    computed_traces = [],              # dendrogram traces from another (precomputed) Clustergram component    \n",
    "#    row_colors = [],                   # list of row colors (observation colors)\n",
    "#    row_colors_label = '',             # string which describes the annotation label for row_colors\n",
    "#    column_colors = [],                # list of column colors (observation colors)\n",
    "#    column_colors_label = '',          # string which describes the annotation label for column_colors\n",
    "#    hidden_labels = [],                # optional list with elements 'row' and/or 'col' to hide corresponding labels on the final plot     \n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3f0fe40-5004-48b3-a248-b983a16f2e5f",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 4. Update config for built-in options for the interactivity\n",
    "At the top right of the interactive chart, when you hover over it, a manu appears with pre-built options for interactivity. By defining **config**, you can add or exclude certain options and make the chart even more customizable. In particular, *`modeBarButtonsToAdd`* allows you to add additional buttons with options that, for example, allow you to draw lines or shapes on the graph. This way you can highlight important points  or parts of the chart."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e53ce45-d6cb-4324-8300-b857c23f697c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "config = {\n",
    "    'responsive': True, 'showTips': True,\n",
    "    'modeBarButtonsToAdd':['drawline', 'drawopenpath', 'drawclosedpath', 'drawcircle', 'drawrect',\n",
    "                           'eraseshape', 'resetViews', 'toggleHover', 'toggleSpikelines'], \n",
    "    'toImageButtonOptions': {\n",
    "        'format': static_img_format,        # one of png, svg, jpeg, webp\n",
    "        'filename': static_img_filename,    # enter the default name for static image available for download\n",
    "        'height': static_img_height,        # dimensions of the exported static image can be different than for the chart available in this notebook\n",
    "        'width': static_img_width,          \n",
    "        'scale': static_img_scale,          # multiply font size for the title/legend/axis/canvas sizes by this factor\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42e6dc94-b3a9-4d4d-9cce-0ce3502034b4",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 5. Update graph layout settings\n",
    "Figure object is composed of two components:\n",
    "* data\n",
    "* layout\n",
    "\n",
    "You can customize a number of **layout** elements by calling the *`update_layout()`* function. In this example, you can learn the most important attributes. However, there are many more, which you can learn in detail from the [Plotly documentation](https://plotly.com/python/reference/layout/).\n",
    "\n",
    "There are also other *update* functions, such as *`update_traces()`*, which allows you to modify the settings for a particular trace (**data**), especially if there are more than one in the chart. In this example, we have modified the point markers by adding a black border to the colored points and adjusting their size.\n",
    "\n",
    "<span style=\"color:gray\">*Enjoy exploring interactivity in the graph.*</span>\n",
    "* <span style=\"color:gray\">*Preview the annotation of a point with interactive labels that appear when you hover over the point.*</span>\n",
    "* <span style=\"color:gray\">*Select a region of the graph to zoom in on it.*</span>\n",
    "* <span style=\"color:gray\">*Manage the display of the data series by clicking on the 'state code' in the legend (on the right).*</span>\n",
    "* <span style=\"color:gray\">*Use the options in the built-in mode bar at the top of the plotting area, to add additional objects to the chart.*</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae75ea47-969b-48fb-9678-bf674d802f13",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adjust heatmap\n",
    "xtick_vals = []\n",
    "ytick_vals = []\n",
    "ytick_labs = []\n",
    "y_step = 1\n",
    "x_range = []\n",
    "row_clusters = {}\n",
    "col_clusters = {}\n",
    "\n",
    "for trace in fig[\"data\"]:\n",
    "    if trace.type == 'heatmap':\n",
    "        yticks_dict = dict(zip(trace.y, list(df[y_col_name])))\n",
    "        if y_ticks_str:            # prepare text labels for Y axis ticks \n",
    "            arr = np.asarray(list(yticks_dict.keys()))\n",
    "            if y_zoom[0] == 0:\n",
    "                y_zoom[0] = min(trace.y)\n",
    "            else:\n",
    "                val = round(y_zoom[0] * max(trace.y),1)\n",
    "                y_zoom[0] = arr[(np.abs(arr - val)).argmin()]\n",
    "            if y_zoom[1] == 1:\n",
    "                y_zoom[1] = max(trace.y)\n",
    "            else:\n",
    "                val = round(y_zoom[1] * max(trace.y),1)\n",
    "                y_zoom[1] = arr[(np.abs(arr - val)).argmin()]\n",
    "            y_step = (y_zoom[1] - y_zoom[0])/y_nticks\n",
    "            val = y_zoom[0]\n",
    "            for i in range(0,y_nticks+1):\n",
    "                z = arr[(np.abs(arr - val)).argmin()]\n",
    "                ytick_vals.append(z)\n",
    "                ytick_labs.append(yticks_dict[z])\n",
    "                val += y_step\n",
    "        else:                      # else, use numerical Y axis indices for ticks' labels\n",
    "            y_step = trace.y[1] - trace.y[0]\n",
    "            y_zoom = [min(trace.y) - y_step, max(trace.y) + y_step]\n",
    "        x_step = (max(trace.x) - min(trace.x))/17\n",
    "        x_range = [min(trace.x) - x_step, max(trace.x) + x_step]\n",
    "        xtick_vals = trace.x\n",
    "        # Adjust other attributes of the Heatmap section and used colorscale\n",
    "        trace.xaxis = 'x1'\n",
    "        trace.yaxis = 'y1'\n",
    "        trace.showscale = show_colorbar\n",
    "        trace.colorbar.len = colorbar_len \n",
    "        trace.colorbar.x = colorbar_x\n",
    "        trace.colorbar.y = colorbar_y\n",
    "\n",
    "        # adjust the content of the Heatmap interactive label\n",
    "        trace.name = heatmap_name \n",
    "        if h_label_x != 'none' or h_label_x != '':\n",
    "            trace.hovertemplate = h_label_x + ': %{x}<br>'\n",
    "        if h_label_y != 'none' or h_label_y != '':\n",
    "            trace.text = np.tile(list(df[y_col_name]), (len(trace.x), 1))\n",
    "            trace.text = np.transpose(trace.text)\n",
    "            trace.hovertemplate += h_label_y + ': %{text}<br>'\n",
    "        if h_label_z != 'none' or h_label_z != '':\n",
    "            trace.hovertemplate += h_label_z + ': %{z}<br>'\n",
    "        for num, lab_var in enumerate([h_label_4, h_label_5]):\n",
    "            if num == 0:\n",
    "                attr = \"hovertext\"\n",
    "            else:\n",
    "                attr = \"customdata\"\n",
    "            if lab_var != 'none':\n",
    "                if isinstance(lab_var[2], str):                       # column name\n",
    "                    if lab_var[3] != '' and lab_var[3] != 'none':   # data from external file\n",
    "                        data_4 = pd.read_csv(str(lab_var[2]), sep=col_separator)[lab_var[2]]\n",
    "                    else:\n",
    "                        data_4 = df[lab_var[2]]\n",
    "                else:\n",
    "                    data_4 = lab_var[2]\n",
    "                if lab_var[1] == 'z':\n",
    "                    setattr(trace, attr, data_4)\n",
    "                elif lab_var[1] == 'x':\n",
    "                    setattr(trace, attr, np.tile(data_4, (len(trace.y), 1)))\n",
    "                elif h_label_4[1] == 'y':\n",
    "                    tmp = np.tile(data_4, (len(trace.x), 1))\n",
    "                    setattr(trace, attr, np.transpose(tmp))\n",
    "                trace.hovertemplate += lab_var[0] + ': %{'+attr+'}<br>' + '<extra><b>' + trace.name + '</b></extra>'\n",
    "                \n",
    "    # Derive data from generic dendrogram clusters (ranges, trace color)\n",
    "    elif trace.type == 'scatter':\n",
    "        if trace.name != None:\n",
    "            if trace.name.startswith(\"Row\"):\n",
    "                trace.text = trace.name.replace('Row Cluster ', '').strip()                # stores cluster ID (index)\n",
    "                row_clusters[trace.text] = [trace.y[0], trace.y[3], '', trace.marker['color'], trace.x[1]]\n",
    "            elif trace.name.startswith(\"Col\"):\n",
    "                trace.text = trace.name.replace('Col Cluster ', '').strip()                # stores cluster ID (index)\n",
    "                col_clusters[trace.text] = [trace.x[0], trace.x[3], '', trace.marker['color'], trace.y[1]]\n",
    "print('Heatmap trace adjusted.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9298aa4c-3959-4695-88e6-ee32d47087cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adjust dendrogram traces and generate data about children clusters\n",
    "for idx, trace in enumerate(fig[\"data\"]):\n",
    "    if trace.type == 'scatter':\n",
    "        if trace.name != None:                 # dendrograms\n",
    "            trace.hovertext = ''\n",
    "            trace.hoverinfo = 'text + name'\n",
    "            if trace.name.startswith(\"Row\"):   # Y-dendrograms\n",
    "                if row_prefix != '':\n",
    "                    trace.name = trace.name.replace('Row Cluster', row_prefix).strip()         # custom prefix name for clusters\n",
    "                trace.xaxis = 'x3'\n",
    "                trace.yaxis = 'y1'\n",
    "                if trace.y[0] in yticks_dict and trace.y[3] in yticks_dict:\n",
    "                    trace.hovertext = str(yticks_dict[trace.y[0]]) + \" and \" + str(yticks_dict[trace.y[3]])\n",
    "                    row_clusters[trace.text][2] = trace.hovertext.replace(' and', ',')\n",
    "                else:\n",
    "                    tmp = ''\n",
    "                    for c in row_clusters:\n",
    "                        if c != trace.text:\n",
    "                            diff = row_clusters[c][1] - (row_clusters[c][1] - row_clusters[c][0])/2\n",
    "                            if trace.y[3] == diff:   \n",
    "                                trace.hovertext += \"cluster \" + c + \" + \"\n",
    "                                if trace.y[0] in yticks_dict:\n",
    "                                    tmp += str(yticks_dict[trace.y[0]]) + \" + \"\n",
    "                            elif trace.y[0] == diff:\n",
    "                                trace.hovertext += \"cluster \" + c + \" + \"\n",
    "                                if trace.y[3] in yticks_dict:\n",
    "                                    tmp += str(yticks_dict[trace.y[3]]) + \" + \"\n",
    "                    trace.hovertext = trace.hovertext.rstrip('+ ')\n",
    "                    # Check if cluster has only 2 children\n",
    "                    if len(trace.hovertext.split(\" + \")) < 2 and len(tmp.split(\"+\"))-1 > 0:\n",
    "                        trace.hovertext +=  ' and ' + tmp.rstrip('+ ')\n",
    "                    if len(trace.hovertext.split(\" + \")) > 2:\n",
    "                        cls = [i.replace('cluster', '').strip() for i in trace.hovertext.split('+')] # cluster ids\n",
    "                        rls = [i for i in cls if row_clusters[i][4] < trace.x[1]] # cluster x position\n",
    "                        cls = [i for i in cls if i not in rls]\n",
    "                        while len(cls) > 2:\n",
    "                            vls = [row_clusters[i][4] for i in cls] # cluster x position\n",
    "                            idx = (np.abs(np.asarray(vls) - trace.x[1])).argmax()\n",
    "                            cls.pop(idx)\n",
    "                        trace.hovertext = \"cluster \" + cls[0] + \" + \" + \"cluster \" + cls[1]\n",
    "                    row_clusters[trace.text][2] = trace.hovertext.replace(' and', ',').replace(' +', ',')                                  \n",
    "                trace.hovertext = '<b>' + trace.name + ' merges:</b><br>' + trace.hovertext\n",
    "                trace.hovertemplate = trace.hovertext + '<extra><b>dendrogram Y</b><br>' + trace.name + '</extra>'\n",
    "\n",
    "            elif trace.name.startswith(\"Col\"):   # X-dendrograms\n",
    "                if row_prefix != '':\n",
    "                    trace.name = trace.name.replace('Col Cluster', col_prefix)\n",
    "                trace.xaxis = 'x1'\n",
    "                trace.yaxis = 'y3'\n",
    "                xticks_dict = dict(zip(xtick_vals, xtick_labs))\n",
    "                if trace.x[0] in xticks_dict and trace.x[3] in xticks_dict:\n",
    "                    trace.hovertext = str(xticks_dict[trace.x[0]]) + \" and \" + str(xticks_dict[trace.x[3]])\n",
    "                    col_clusters[trace.text][2] = trace.hovertext.replace(' and', ',')                    \n",
    "                else:\n",
    "                    tmp = ''\n",
    "                    for c in col_clusters:\n",
    "                        if c != trace.text:\n",
    "                            diff = col_clusters[c][1] - (col_clusters[c][1] - col_clusters[c][0])/2\n",
    "                            if trace.x[3] == diff:   \n",
    "                                trace.hovertext += \"cluster \" + c + \" + \"\n",
    "                                if trace.x[0] in xticks_dict:\n",
    "                                    tmp += str(xticks_dict[trace.x[0]]) + \" + \"\n",
    "                            elif trace.x[0] == diff:\n",
    "                                trace.hovertext += \"cluster \" + c + \" + \"\n",
    "                                if trace.x[3] in xticks_dict:\n",
    "                                    tmp += str(xticks_dict[trace.x[3]]) + \" + \"\n",
    "                    trace.hovertext = trace.hovertext.rstrip('+ ')\n",
    "                    # Check if cluster has only 2 children\n",
    "                    if len(trace.hovertext.split(\" + \")) < 2 and len(tmp.split(\"+\"))-1 > 0:\n",
    "                        trace.hovertext +=  ' and ' + tmp.rstrip('+ ')\n",
    "                    if len(trace.hovertext.split(\" + \")) > 2:\n",
    "                        cls = [i.replace('cluster', '').strip() for i in trace.hovertext.split('+')] # cluster ids\n",
    "                        rls = [i for i in cls if col_clusters[i][4] < trace.y[1]] # cluster x position\n",
    "                        cls = [i for i in cls if i not in rls]\n",
    "                        while len(cls) > 2:\n",
    "                            vls = [col_clusters[i][4] for i in cls] # cluster x position\n",
    "                            idx = (np.abs(np.asarray(vls) - trace.y[1])).argmax()\n",
    "                            cls.pop(idx)\n",
    "                        trace.hovertext = \"cluster \" + cls[0] + \" + \" + \"cluster \" + cls[1]\n",
    "                    col_clusters[trace.text][2] = trace.hovertext.replace(' and', ',').replace(' +', ',')\n",
    "                trace.hovertext = '<b>' + trace.name + ' merges:</b><br>' + trace.hovertext\n",
    "                trace.hovertemplate = trace.hovertext + '<extra><b>dendrogram X</b><br>' + trace.name + '</extra>'\n",
    "print(\"Dendrogram data prepared.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e2902eb-713d-41b2-8f89-b1ed89c82db8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare data for Y-dendrogram interactive labels and ranges for Y-bars\n",
    "row_content = {}                # all elements for a given cluster\n",
    "cluster_range = {}              # real ranges for a given cluster\n",
    "for c in row_clusters.keys():\n",
    "    row_content[c] = row_clusters[c][2]\n",
    "    cluster_range[c] = [row_clusters[c][0], row_clusters[c][1]]\n",
    "    tokens = [j.strip() for j in row_clusters[c][2].split(',')]\n",
    "    to_remove = []\n",
    "    for i in tokens:\n",
    "        if i.startswith('cluster'):\n",
    "            c_id = i.replace('cluster', '').strip()\n",
    "            news = [j.strip() for j in row_clusters[c_id][2].split(',')]\n",
    "            tokens.extend(news)\n",
    "            to_remove.append(i)\n",
    "    for j in to_remove:\n",
    "        tokens.remove(j)\n",
    "    ranges = sum([[key for key, val in yticks_dict.items() if val == element] for element in tokens], [])\n",
    "    cluster_range[c] = [min(ranges) - (0.03*y_step), max(ranges) + (0.03*y_step)]\n",
    "    tokens = [i+\", \" for i in tokens]\n",
    "    if len(tokens) > n_per_line_row:\n",
    "        i = n_per_line_row - 1\n",
    "        while i < len(tokens):\n",
    "            tokens[i] += '<br>'\n",
    "            i += n_per_line_row\n",
    "    row_content[c] = str(''.join(tokens)).strip().rstrip(',').rstrip(', <br>')\n",
    "\n",
    "print(\"Y-dendrogram interactive labels prepared.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24aad919-3888-41e4-9020-89bd76d92ac3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare data for X-dendrogram interactive labels and ranges for X-bars\n",
    "col_range = {}              # real ranges for a given cluster\n",
    "for c in col_clusters.keys():\n",
    "    col_range[c] = [col_clusters[c][0], col_clusters[c][1]]\n",
    "    tokens = [j.strip() for j in col_clusters[c][2].split(',')]\n",
    "    to_remove = []\n",
    "    for i in tokens:\n",
    "        if i.startswith('cluster'):\n",
    "            c_id = i.replace('cluster', '').strip()\n",
    "            news = [j.strip() for j in col_clusters[c_id][2].split(',')]\n",
    "            tokens.extend(news)\n",
    "            to_remove.append(i)\n",
    "    for j in to_remove:\n",
    "        tokens.remove(j)\n",
    "    ranges = sum([[key for key, val in xticks_dict.items() if val == element] for element in tokens], [])\n",
    "    col_range[c] = [min(ranges) - (0.01*y_step), max(ranges) + (0.01*y_step)]\n",
    "\n",
    "print(\"X-dendrogram interactive labels prepared.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c003f07-e8a0-412e-aa6e-97258577dbb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adjust dendrograms and bars traces\n",
    "for idx, trace in enumerate(fig[\"data\"]):\n",
    "    if trace.type == 'scatter':\n",
    "        # List all elements grouped in a given cluster (Y dendrogram)\n",
    "        if trace.name != None and list_all_row:\n",
    "            if trace.name.startswith(\"Row\") or trace.name.startswith(row_prefix):    # Y dendrogram\n",
    "                trace.hovertext += '<br><br>' + \"cluster size: \" + str(len(row_content[trace.text].split(','))) + \"<br>\" + row_content[trace.text]\n",
    "                trace.hovertemplate = trace.hovertext + '<extra><b>dendrogram Y</b><br>' + trace.name + '</extra>'\n",
    "\n",
    "        # Adjust cluster bars (update real length and trace color)\n",
    "        else:\n",
    "            if trace.x == (0,0):                 # Y-bars\n",
    "                trace.xaxis = 'x2'\n",
    "                trace.yaxis = 'y1'\n",
    "                trace.hoverinfo = 'text+name'\n",
    "                for i in row_markers:\n",
    "                    gid = str(i['group'])\n",
    "                    if list(trace.y) == [row_clusters[gid][0], row_clusters[gid][1]]:\n",
    "                        trace.name = \"merged cluster \" + gid\n",
    "                        trace.y = cluster_range[gid]\n",
    "                        trace.text = '<b>merged cluster: ' + gid + '</b><br>' + str(i['annotation'])\n",
    "                        if i['color'] == 'white':\n",
    "                            trace.line['color'] = row_clusters[gid][3]\n",
    "                trace.hovertemplate = trace.text + '<extra><b>bar cluster Y</b><br>' + trace.name + '</extra>'\n",
    "            \n",
    "            elif trace.y == (0,0):               # X-bars\n",
    "                trace.xaxis = 'x1'\n",
    "                trace.yaxis = 'y2'\n",
    "                trace.hoverinfo = 'text+name'\n",
    "                for i in col_markers:\n",
    "                    gid = str(i['group'])\n",
    "                    if list(trace.x) == [col_clusters[gid][0], col_clusters[gid][1]]:\n",
    "                        trace.name = \"merged cluster \" + gid\n",
    "                        trace.x = col_range[gid]\n",
    "                        trace.text = '<b>merged cluster: ' + gid + '</b><br>' + str(i['annotation'])\n",
    "                        if i['color'] == 'white':\n",
    "                            trace.line['color'] = col_clusters[gid][3]\n",
    "                trace.hovertemplate = trace.text + '<extra><b>bar cluster X</b><br>' + trace.name + '</extra>'\n",
    "                \n",
    "print(\"Adjusted dendrograms and bars traces.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "832d2a96-ab3c-41f9-8c11-e0d6c6944be1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Clean Layout from redundant axis objects and annotations\n",
    "fig.layout.annotations = []\n",
    "for i in fig.layout:\n",
    "    if i.startswith('xaxis'):\n",
    "        fig.layout.pop(i, None)\n",
    "    if i.startswith('yaxis'):\n",
    "        fig.layout.pop(i, None)\n",
    "\n",
    "# Prepare margins\n",
    "graph_margins = {}\n",
    "if margins[0] != 'd':\n",
    "    graph_margins['l'] = margins[0]  \n",
    "if margins[1] != 'd':\n",
    "    graph_margins['t'] = margins[1]\n",
    "if margins[2] != 'd':\n",
    "    graph_margins['r'] = margins[2] \n",
    "if margins[3] != 'd':\n",
    "    graph_margins['b'] = margins[3] \n",
    "        \n",
    "# Update the Layout settings\n",
    "fig.update_layout(\n",
    "    # General settings\n",
    "    height=graph_height, width=graph_width, margin=graph_margins,\n",
    "    paper_bgcolor=graph_bg_color,\n",
    "    plot_bgcolor=drawing_bg_color,\n",
    "    font=dict(family=font_family),\n",
    "    # Title, Legend, Axes\n",
    "    title=dict(text=graph_title, font=dict(color=ti_color, size=ti_size), x=ti_x, y=ti_y),\n",
    "    showlegend=is_legend, legend = dict(x=legend_x, y=legend_y),\n",
    "    xaxis1=dict(title=dict(text=x_title, font=dict(color=x_ti_color, size=x_ti_size)),                                  # heatmap, horizontal (cols): bars, dendro\n",
    "                range = x_range, domain=[dendro_ratio[1],1], side=x_tick_lab_side,\n",
    "                showline=x_axis_show, linewidth=x_axis_width, linecolor=x_axis_color, mirror=x_axis_mirror, \n",
    "                ticks=x_axis_ticks, tickwidth=x_tick_width, ticklen=x_tick_len, tickcolor=x_tick_color,\n",
    "                showticklabels=show_xtick_lab, tickangle=x_tick_lab_angle, \n",
    "                tickmode='array', tickvals=xtick_vals, ticktext=xtick_labs, \n",
    "                tickfont=dict(color=x_tick_lab_color, size=x_tick_lab_size),\n",
    "               ),\n",
    "    xaxis2=dict(domain=[dendro_ratio[1]*dendro_factor,dendro_ratio[1]], showticklabels=False),                          # vertical bars\n",
    "    xaxis3=dict(domain=[0,dendro_ratio[1]*dendro_factor], showticklabels=False),                                        # vertical dendro\n",
    "    yaxis1=dict(title=dict(text=\"\", font=dict(color=y_ti_color, size=y_ti_size)),                                       # heatmap, vertical (rows): bars, dendro\n",
    "                range=y_zoom, domain=[0,1-dendro_ratio[0]], side=y_tick_lab_side,\n",
    "                showline=y_axis_show, linewidth=y_axis_width, linecolor=y_axis_color, mirror=y_axis_mirror,\n",
    "                ticks=y_axis_ticks, tickwidth=y_tick_width, ticklen=y_tick_len, tickcolor=y_tick_color,\n",
    "                showticklabels=show_ytick_lab, tickangle=y_tick_lab_angle, fixedrange=False,\n",
    "                tickfont=dict(color=y_tick_lab_color, size=y_tick_lab_size),\n",
    "               ),\n",
    "    yaxis2=dict(domain=[1-dendro_ratio[0],1-dendro_ratio[0]*dendro_factor*0.8], showticklabels=False),                  # horizontal bars\n",
    "    yaxis3=dict(domain=[1-dendro_ratio[0]*dendro_factor*0.8,1], showticklabels=False),                                  # horizontal dendro\n",
    "    # Interactive hover labels\n",
    "    hoverlabel=dict(namelength = -1, align=hover_align, font=dict(size=hover_size)),\n",
    ")\n",
    "\n",
    "# Update Heatmap Y axis ticks if 'text' labels are provided\n",
    "if y_ticks_str:\n",
    "    fig.update_layout(\n",
    "        yaxis1=dict(tickmode = 'array', tickvals = ytick_vals, ticktext = ytick_labs, fixedrange = True, range=y_zoom,) \n",
    "    )\n",
    "    \n",
    "fig.add_annotation(x=y_ti_x, y=y_ti_y, showarrow=False, xref='paper', yref='paper',                                     # customized title of Y axis\n",
    "        text=y_title, textangle=y_ti_angle, font=dict(color=y_ti_color, size=y_ti_size),\n",
    ")\n",
    "\n",
    "\n",
    "fig.show(config=config)              # to display graph interally in the notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fda46e7-636f-4523-b01c-1463656cfe98",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 6. Export interactive HTML chart and/or Write static image file\n",
    "A refined chart can be exported in many different ways. \n",
    "* First, you can use the built-in options bar, located directly above the interactive chart. There, using the camera icon, you can download a static image in the format and size defined in config.\n",
    "* You can also export a static image in any format (vector or raster) directly from the command line (see the code cell below).\n",
    "* Also, you can export the interactive chart as HTML and even have it open immediately in a new browser tab.\n",
    "\n",
    "*<span style=\"color:gray\">If you want to export the image, remove the **'#'** tag in the selected row of code below. <br>Then run the cell again (`ctrl+enter`). <br>If you observe an error, you have probably not installed the kaleido package yet. Quick fix: `pip install -U kaleido`.</span>*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d72d90ef-6a76-4637-a97e-4331143bdeb1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "print(\"Static image saved to: \" + static_img_filename + '.' + static_img_format)\n",
    "fig.write_image(static_img_filename + '.' + static_img_format)          # static image exported in the selected format\n",
    "fig.write_html(static_img_filename + '.html', auto_open=html_auto)      # interactive HTML chart (opens automatically in the new browser tab)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ca23867-9e69-4658-895a-57273e5596b7",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Static image of the graph\n",
    "\n",
    "*(Please edit this markdown cell and provide the name of the desired static image file.)*\n",
    "\n",
    "![PC1 vs PC2](example_clustergram.png)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
